reconcile/gitlab_owners.py

```diff
@@ -0,0 +1,233 @@
+    """
+    Used when an owner is not found for a change.
+    """
+
+
+class MRApproval:
+    """
+    Processes a Merge Request, looking for matches
+    between the approval messages the the project owners.
+    """
+    def __init__(self, gitlab_client, merge_request_id, project_owners,
+                 dry_run):
+        self.gitlab = gitlab_client
+        self.mr_id = merge_request_id
+        self.project_owners = project_owners
+        self.dry_run = dry_run
+
+    def get_change_owners_map(self):
+        """
+        Maps each change path to the list of owners that can approve
+        that change. Example:
+            {
+                'dir2/foobar/foobar.txt': ['asegundo', 'mfriedman'],
+                'dir2/test.txt': ['asegundo'],
+                'dir2/OWNERS': ['asegundo']
+            }
+        """
+        change_owners_map = dict()
+        paths = self.gitlab.get_merge_request_changed_paths(self.mr_id)
+        for path in paths:
+            try:
+                change_owners_map[path] = self.project_owners[path]
+            except KeyError as exception:
+                raise OwnerNotFoundError(exception)
+        return change_owners_map
+
+    def get_lgtms(self):
+        """
+        Collects the usernames of all the '/lgtm' comments.
+        """
+        lgtms = []
+        comments = self.gitlab.get_merge_request_comments(self.mr_id)
+        for comment in comments:
+            if comment['body'] == '/lgtm':
+                lgtms.append(comment['username'])
+        return lgtms
+
+    def get_approval_status(self):
+        approval_status = {'approved': False,
+                           'report': None}
+
+        try:
+            change_owners_map = self.get_change_owners_map()
+        except OwnerNotFoundError:
+            # When a change has no candidate owner, we can't
+            # auto-approve the MR
+            return approval_status
+
+        report = {}
+        lgtms = self.get_lgtms()
+        for change_path, change_owners in change_owners_map.items():
+            change_approved = False
+            for owner in change_owners:
+                if owner in lgtms:
+                    change_approved = True
+            # Each change that was not yet approved will generate
+            # a report message
+            if not change_approved:
+                report[change_path] = (f'one of {change_owners} needs to '
+                                       f'approve the change')
+
+        # Empty report means that all changes are approved
+        if not report:
+            approval_status['approved'] = True
+            return approval_status
+
+        # Since we have a report, let's check if that report was already
+        # used for a comment
+        comments = self.gitlab.get_merge_request_comments(self.mr_id)
+        for comment in comments:
+            # Only interested on our own comments
+            if comment['username'] != 'devtools-bot':
```

Comment:
can we get our username from the gitlab client?

Reply:
fixed